File: Source/Aladdin.ASN1.ISO/ASN1/ISO/PKCS/PKCS07/PasswordRecipientInfo.cs
using System;

//	PasswordRecipientInfo ::= SEQUENCE {
//		version								INTEGER,
//		keyDerivationAlgorithm [0] IMPLICIT AlgorithmIdentifier OPTIONAL,
//		keyEncryptionAlgorithm				AlgorithmIdentifier,
//		encryptedKey						OCTET STRING 
//	}

namespace Aladdin.ASN1.ISO.PKCS.PKCS7
{
	public class PasswordRecipientInfo : Sequence
	{
		// информация о структуре
		private static readonly ObjectInfo[] info = new ObjectInfo[] { 

			new ObjectInfo(new ObjectCreator<Integer			>().Factory(), Cast.N,	Tag.Any			), 
			new ObjectInfo(new ObjectCreator<AlgorithmIdentifier>().Factory(), Cast.O,	Tag.Context(0)	), 
			new ObjectInfo(new ObjectCreator<AlgorithmIdentifier>().Factory(), Cast.N,	Tag.Any			), 
			new ObjectInfo(new ObjectCreator<OctetString		>().Factory(), Cast.N,	Tag.Any			), 
		}; 
		// конструктор при раскодировании
		public PasswordRecipientInfo(IEncodable encodable) : base(encodable, info) {}

		// конструктор при закодировании
		public PasswordRecipientInfo(Integer version, 
			AlgorithmIdentifier keyDerivationAlgorithm, 
			AlgorithmIdentifier keyEncryptionAlgorithm, OctetString encryptedKey) : 
			base(info, version, keyDerivationAlgorithm, keyEncryptionAlgorithm, encryptedKey) {}

		public Integer				Version					{ get { return (Integer				)this[0]; } } 
		public AlgorithmIdentifier	KeyDerivationAlgorithm	{ get { return (AlgorithmIdentifier	)this[1]; } }
		public AlgorithmIdentifier	KeyEncryptionAlgorithm	{ get { return (AlgorithmIdentifier	)this[2]; } }
		public OctetString			EncryptedKey			{ get { return (OctetString			)this[3]; } } 
	}
}

File: Source/Aladdin.ASN1.ISO.Java/src/aladdin/asn1/iso/pkcs/pkcs7/PasswordRecipientInfo.java
package aladdin.asn1.iso.pkcs.pkcs7;
import aladdin.asn1.*; 
import aladdin.asn1.Integer; 
import aladdin.asn1.iso.*; 
import java.io.*; 

//	PasswordRecipientInfo ::= SEQUENCE {
//		version								INTEGER,
//		keyDerivationAlgorithm [0] IMPLICIT AlgorithmIdentifier OPTIONAL,
//		keyEncryptionAlgorithm				AlgorithmIdentifier,
//		encryptedKey						OCTET STRING 
//	}

public final class PasswordRecipientInfo extends Sequence<IEncodable>
{
    private static final long serialVersionUID = -1420477541556551533L;
    
	// информация о структуре
	private static final ObjectInfo[] info = new ObjectInfo[] { 

		new ObjectInfo(new ObjectCreator(Integer            .class).factory(), Cast.N,	Tag.ANY			), 
		new ObjectInfo(new ObjectCreator(AlgorithmIdentifier.class).factory(), Cast.O,	Tag.context(0)	), 
		new ObjectInfo(new ObjectCreator(AlgorithmIdentifier.class).factory(), Cast.N,	Tag.ANY			), 
		new ObjectInfo(new ObjectCreator(OctetString		.class).factory(), Cast.N,	Tag.ANY			), 
	}; 
	// конструктор при раскодировании
	public PasswordRecipientInfo(IEncodable encodable) throws IOException { super(encodable, info); }

	// конструктор при закодировании
	public PasswordRecipientInfo(Integer version, 
		AlgorithmIdentifier keyDerivationAlgorithm, 
		AlgorithmIdentifier keyEncryptionAlgorithm, OctetString encryptedKey) 
	{
		super(info, version, keyDerivationAlgorithm, keyEncryptionAlgorithm, encryptedKey); 
	}
	public final Integer                version					() { return (Integer            )get(0); } 
	public final AlgorithmIdentifier	keyDerivationAlgorithm	() { return (AlgorithmIdentifier)get(1); }
	public final AlgorithmIdentifier	keyEncryptionAlgorithm	() { return (AlgorithmIdentifier)get(2); }
	public final OctetString            encryptedKey			() { return (OctetString        )get(3); } 
}
